### PR TITLE
Optimize endoscopy tool tracking app

### PR DESCRIPTION
**Pull request by AndreasHeumann**
Optimize endoscopy tool tracking app

Modify ToolTrackingPostProcessorOp to generate coordinates using
CUDA. This avoids the roundtrip from GPU to CPU memory and back.

Use RMMAllocator for video replayer.

Fix type  of `dla_core` parameter of TensorRtInference.

Remove unused `clock` parameter.

Use single `out` port instead of `out_coords` and `out_mask` since
both outputs are now in device memory.

Latency measured with flow benchmarking on x86 with RTX 6000 Ada
and replayer `realtime`set to `false`:
before optimization 4.8 ms, after optimization 3.09 ms.

### CODE REVIEW

File: operators/lstm_tensor_rt_inference/lstm_tensor_rt_inference.cpp
Comment (grlee77) [inline operators/lstm_tensor_rt_inference/lstm_tensor_rt_inference.cpp:98]: thanks for adding this to suppress warnings in the logs when it is not provided

File: operators/lstm_tensor_rt_inference/python/lstm_tensor_rt_inference.cpp
Comment (grlee77) [inline operators/lstm_tensor_rt_inference/python/lstm_tensor_rt_inference.cpp:3]:
```suggestion
 * SPDX-FileCopyrightText: Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
```

File: operators/tool_tracking_postprocessor/CMakeLists.txt
Comment (grlee77) [inline operators/tool_tracking_postprocessor/CMakeLists.txt:29]: I see CUDA_ARCHITECTURES was removed. Are these already being set at the top-level of the project?

File: operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp
Comment (grlee77) [inline operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp:3]:
```suggestion
 * SPDX-FileCopyrightText: Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
```

File: operators/tool_tracking_postprocessor/python/tool_tracking_postprocessor.cpp
Comment (grlee77) [inline operators/tool_tracking_postprocessor/python/tool_tracking_postprocessor.cpp:81]: If this is removed, I think you likely also need to update the other variants of the endoscopy tool tracking app that are using this operator:

- `applications/endoscopy_tool_tracking_distributed`
- `applications/h264/h264_endoscopy_tool_tracking`

File: operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp
Comment (grlee77) [inline operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp:193]: Originally these tensors were both on a single "out" port, but I [had to change to separate ports for the distributed version of the app](https://github.com/nvidia-holoscan/holohub/pull/262) because UCX cannot mix host and device tensors in the same entity transmitted by `UcxTransmitter` (the same UCX transport layer, e.g. cuda_copy, etc. has to be used for all tensors in a given entity)
```suggestion
  op_output.emit(result_host, "out_coords");
  op_output.emit(result_device, "out_mask");
```

Now that it looks like `out_coords` and `out_mask` are both in device memory, then it would be good to go back to having only a single "out" port again, which may be a bit more efficient, particularly in the distributed case as it reduces the number of separate messages that have to be sent between the fragments.

File: applications/endoscopy_tool_tracking/cpp/main.cpp
Comment (grlee77) [inline applications/endoscopy_tool_tracking/cpp/main.cpp:166]: Is leaving it as `BlockMemoryPool` instead of `RMMAllocator` here is mainly so we don't have to assign a different allocator depending on the version of Holoscan? (or was there some performance benefit to keeping the block memory pool instead?)

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks, Andreas it looks great. I had two main questions left to consider:

1.) I think a couple other apps using tool tracking will need corresponding changes in the operator arguments (no `host_allocator` in order to continue to work.
2.) Do you want to merge "out_mask" and "out_coords" back to just a single "out" port like they were before #262 was merged

File: operators/tool_tracking_postprocessor/CMakeLists.txt
Comment (AndreasHeumann) [reply operators/tool_tracking_postprocessor/CMakeLists.txt:29]: No, I don't think so. If it is not specified then it's using the default chosen by ncc. Using "70;80" seems to be somewhat artificially used. Using "all" like in HSDK does not work since the x86 CUDA driver does not include the kernels for mobile GPUs (e.g. SM87 (Orin)). Since the compiled code usually is used on the installed GPU using that option seems to make most sense.

File: operators/tool_tracking_postprocessor/python/tool_tracking_postprocessor.cpp
Comment (AndreasHeumann) [reply operators/tool_tracking_postprocessor/python/tool_tracking_postprocessor.cpp:81]: Thanks, I forgot these.

File: operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp
Comment (AndreasHeumann) [reply operators/tool_tracking_postprocessor/tool_tracking_postprocessor.cpp:193]: Yes, makes sense. I'll change that back to a single output port.

File: applications/endoscopy_tool_tracking/cpp/main.cpp
Comment (AndreasHeumann) [reply applications/endoscopy_tool_tracking/cpp/main.cpp:166]: This is just to make the code less cluttered and easier to read. For the video replayer we only have RMM or Unbound allocator, so adding the extra code to use RMM allocator makes sense. But for the others I think using BlockMemory and RMM allocator makes no difference.

Comment (grlee77) [reply applications/endoscopy_tool_tracking/cpp/main.cpp:166]: thanks for confirming. I am fine with this as-is then.

File: operators/tool_tracking_postprocessor/CMakeLists.txt
Comment (tbirdso) [inline operators/tool_tracking_postprocessor/CMakeLists.txt:1]:
```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2023-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

**Review by tbirdso — APPROVED** [review]
Looks good! One copyright fixup. Will wait for approval from @grlee77 before merge.

**Review by grlee77 — APPROVED** [review]
Thanks for making the suggested changes, approved!

File: operators/tool_tracking_postprocessor/CMakeLists.txt
Comment (AndreasHeumann) [reply operators/tool_tracking_postprocessor/CMakeLists.txt:1]: 2034 :-)

**Comment by AndreasHeumann** [comment]
Thanks, fixed the copyright year. Ready to be merged.

**Pull request merged**
